pallets/pool-registry/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+// Copyright 2022 Centrifuge Foundation (centrifuge.io).
+// This file is part of Centrifuge chain project.
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use cfg_primitives::Moment;
+use cfg_traits::Permissions;
+use cfg_types::{PermissionScope, PoolChanges, PoolRole, Role, TrancheInput};
+use codec::HasCompact;
+use frame_support::{pallet_prelude::*, scale_info::TypeInfo, transactional, BoundedVec};
+use frame_system::pallet_prelude::*;
+pub use pallet::*;
+use sp_runtime::{
+	traits::{AtLeast32BitUnsigned, BadOrigin},
+	FixedPointNumber, FixedPointOperand,
+};
+use sp_std::vec::Vec;
+pub use weights::WeightInfo;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+pub mod weights;
+
+#[derive(Debug, Encode, PartialEq, Eq, Decode, Clone, TypeInfo)]
+pub struct TrancheMetadata<MaxTokenNameLength, MaxTokenSymbolLength>
+where
+	MaxTokenNameLength: Get<u32>,
+	MaxTokenSymbolLength: Get<u32>,
+{
+	pub token_name: BoundedVec<u8, MaxTokenNameLength>,
+	pub token_symbol: BoundedVec<u8, MaxTokenSymbolLength>,
+}
+
+#[derive(Clone, Encode, Decode, Eq, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct PoolMetadata<MetaSize>
+where
+	MetaSize: Get<u32>,
+{
+	metadata: BoundedVec<u8, MetaSize>,
+}
+
+pub trait PoolMutate<
+	AccountId,
+	Balance,
+	PoolId,
+	CurrencyId,
+	Rate,
+	MaxTokenNameLength: Get<u32>,
+	MaxTokenSymbolLength: Get<u32>,
+	MaxTranches: Get<u32>,
+>
+{
+	fn create(
+		admin: AccountId,
+		depositor: AccountId,
+		pool_id: PoolId,
+		tranche_inputs: Vec<TrancheInput<Rate, MaxTokenNameLength, MaxTokenSymbolLength>>,
+		currency: CurrencyId,
+		max_reserve: Balance,
+		metadata: Option<Vec<u8>>,
+	) -> DispatchResult;
+
+	fn update(
+		pool_id: PoolId,
+		changes: PoolChanges<Rate, MaxTokenNameLength, MaxTokenSymbolLength, MaxTranches>,
+	) -> DispatchResultWithPostInfo;
+
+	fn execute_update(pool_id: PoolId) -> DispatchResultWithPostInfo;
```

Comment:
PoolSystem will implement this. Move to libs/traits.

Reply:
We will get a circular dependency when moving this to `cfg-traits`, because we rely on `cfg-types` here because of `PoolChanges` and `TrancheInput`. Looking at this, it would make sense not to have any `impls` in `cfg-types` and move them to `cfg-traits`. So `cfg-types` doesn't rely on `cfg-traits`.